hedy.py

```diff
@@ -1310,29 +1301,26 @@ def turn(self, args):
         if arg.lstrip("-").isnumeric():
             return self.make_turn(arg)
 
-    def punctuation(self, args):
-        return ''.join([str(c) for c in args])
     def var(self, args):
         name = args[0]
         self.check_var_usage(args)
         return escape_var(name)
     def var_access(self, args):
         name = args[0]
         return escape_var(name)
-    def print(self, args):
-        argument_string = ""
-        i = 0
 
-        for argument in args:
-            # final argument and punctuation arguments do not have to be separated with a space, other do
-            if i == len(args)-1 or args[i+1] in self.punctuation_symbols:
-                space = ''
+    def print(self, args):
+        args_new = []
+        for a in args:
+            # list access has been already rewritten since it occurs lower in the tree
+            # therefore we should not split it anymore and thread it as 1 variable:
+            if "random.choice" in a or "[" in a:
+                args_new.append(self.process_variable_for_fstring(a))
             else:
-                space = " "
-
-            argument_string += self.process_variable_for_fstring(argument) + space
+                res = regex.findall(r"[\p{Lu}\p{Ll}\p{Lt}\p{Lm}\p{Lo}\p{Nl}\p{Mn}\p{Mc}\p{Nd}\p{Pc}]+|[^\p{Lu}\p{Ll}\p{Lt}\p{Lm}\p{Lo}\p{Nl}]+", a)
```

Comment:
This is a nice, creative approach I would not have thought of!

Reply:
Yeah, this is really smart! And simplifies the grammar!